[llvm][ARM]Add  widen global arrays pass

nasherm opened the pull request:
- Pass optimizes memcpy's by padding out destinations and sources to a full word to make backend generate full word loads instead of loading a single byte (ldrb) and/or half word (ldrh). Only pads destination when it's a stack allocated constant size array and source when it's constant array. Heuristic to decide whether to pad or not is very basic and could be improved to allow more examples to be padded.
- Pass works within GlobalOpt but is disabled by default on all targets except ARM.

nasherm committed:
[llvm][ARM]Add ARM widen strings pass

- Pass optimizes memcpy's by padding out destinations and sources to a
  full word to make ARM backend generate full word loads instead of
  loading a single byte (ldrb) and/or half word (ldrh). Only pads
  destination when it's a stack allocated constant size array and source
  when it's constant string. Heuristic to decide whether to pad or not
  is very basic and could be improved to allow more examples to be
  padded.
- Pass works at the midend level

Change-Id: I1c6371f0962e7ad3c166602b800d041ac1cc7b04

aeubanks commented:
if this is actually arm-specific, please use the `registerPassBuilderCallbacks` framework

some performance numbers in the description would be helpful

adding some arm people

aeubanks commented:
also I would split PRs to implement the pass and add it to the pipeline

davemgreen commented:
I believe there was talk a long time ago about adding this to an existing pass such as the GlobalOpts pass or CGP. It sounds like CGP is too late for it, could it be a part of GlobalOpt or some other pass?

nasherm commented:
> if this is actually arm-specific, please use the `registerPassBuilderCallbacks` framework
> 
> some performance numbers in the description would be helpful
> 
> adding some arm people

I intend to rework the patch to make use of this.

> also I would split PRs to implement the pass and add it to the pipeline

Sure, no problem

nasherm commented:
I've reduced this patch down to adding the pass, as well as tests, without enabling it.

With respect to performance gain I've seen a jump of around 1% on some of our benchmarks.

I used the following (truncated) IR to show the difference in generated assembly
```
# example.ll
@.str = private unnamed_addr constant [10 x i8] c"123456789\00", align 1

; Function Attrs: nounwind
define hidden void @foo() #0 {
entry:
  %something = alloca [10 x i8], align 1
  %arraydecay = getelementptr inbounds [10 x i8], ptr %something, i32 0, i32 0
  %call = call ptr @strcpy(ptr %arraydecay, ptr @.str)
  %arraydecay1 = getelementptr inbounds [10 x i8], ptr %something, i32 0, i32 0
  %call2 = call i32 @bar(ptr %arraydecay1)
  ret void
```
**Optimization off**
```
$ opt example.ll -O2 -S | llc -mtriple=arm-arm-none-eabi -o -
..........
foo:
	.fnstart
@ %bb.0:                                @ %entry
	.save	{r4, lr}
	push	{r4, lr}
	.pad	#24
	sub	sp, sp, #24
	ldr	r12, .LCPI0_0
	add	r0, sp, #4
	mov	lr, r0
	ldm	r12!, {r1, r2, r3, r4}
	stm	lr!, {r1, r2, r3, r4}
	ldrb	r1, [r12]
	strb	r1, [lr]
	bl	bar
	add	sp, sp, #24
	pop	{r4, lr}
	mov	pc, lr
	.p2align	2
@ %bb.1:
.LCPI0_0:
	.long	.L.str
.Lfunc_end0:
	.size	foo, .Lfunc_end0-foo
	.fnend
                                        @ -- End function
	.type	.L.str,%object                  @ @.str
	.section	.rodata.str1.4,"aMS",%progbits,1
	.p2align	2, 0x0
.L.str:
	.asciz	"1234567891234567"
	.size	.L.str, 17

	.section	".note.GNU-stack","",%progbits
	.eabi_attribute	30, 1	@ Tag_ABI_optimization_goals
```

**Optmization on**
```
$ opt example.ll  -passes="default<O2>,arm-widen-strings" -S | llc -mtriple=arm-arm-none-eabi -o -
foo:
	.fnstart
@ %bb.0:                                @ %entry
	.save	{r4, r5, r11, lr}
	push	{r4, r5, r11, lr}
	.pad	#40
	sub	sp, sp, #40
	ldr	r12, .LCPI0_0
	add	r0, sp, #20
	mov	r2, r0
	ldm	r12, {r1, r3, r4, r5, lr}
	stm	r2, {r1, r3, r4, r5, lr}
	bl	bar
	add	sp, sp, #40
	pop	{r4, r5, r11, lr}
	mov	pc, lr
	.p2align	2
@ %bb.1:
.LCPI0_0:
	.long	.L.str
.Lfunc_end0:
	.size	foo, .Lfunc_end0-foo
	.fnend
                                        @ -- End function
	.type	.L__unnamed_1,%object           @ @0
	.section	.rodata.str1.1,"aMS",%progbits,1
.L__unnamed_1:
	.asciz	"1234567891234567"
	.size	.L__unnamed_1, 17

	.type	.L.str,%object                  @ @.str
	.section	.rodata,"a",%progbits
	.p2align	2, 0x0
.L.str:
	.asciz	"1234567891234567\000\000\000"
	.size	.L.str, 20

	.section	".note.GNU-stack","",%progbits
	.eabi_attribute	30, 1	@ Tag_ABI_optimization_goals

```

Diff of assembly for readability
```
24,27c24,27
< 	.save	{r4, lr}
< 	push	{r4, lr}
< 	.pad	#24
< 	sub	sp, sp, #24
---
> 	.save	{r4, r5, r11, lr}
> 	push	{r4, r5, r11, lr}
> 	.pad	#40
> 	sub	sp, sp, #40
29,34c29,32
< 	add	r0, sp, #4
< 	mov	lr, r0
< 	ldm	r12!, {r1, r2, r3, r4}
< 	stm	lr!, {r1, r2, r3, r4}
< 	ldrb	r1, [r12]
< 	strb	r1, [lr]
---
> 	add	r0, sp, #20
> 	mov	r2, r0
> 	ldm	r12, {r1, r3, r4, r5, lr}
> 	stm	r2, {r1, r3, r4, r5, lr}
36,37c34,35
< 	add	sp, sp, #24
< 	pop	{r4, lr}
---
> 	add	sp, sp, #40
> 	pop	{r4, r5, r11, lr}
46a45,50
> 	.type	.L__unnamed_1,%object           @ @0
> 	.section	.rodata.str1.1,"aMS",%progbits,1
> .L__unnamed_1:
> 	.asciz	"1234567891234567"
> 	.size	.L__unnamed_1, 17
> 
48c52
< 	.section	.rodata.str1.4,"aMS",%progbits,1
---
> 	.section	.rodata,"a",%progbits
51,52c55,56
< 	.asciz	"1234567891234567"
< 	.size	.L.str, 17
---
> 	.asciz	"1234567891234567\000\000\000"
> 	.size	.L.str, 20

```

nasherm commented:
> Is there some reason we can't pad globals that aren't strings?
I don't think so? But there might be a reason this wasn't investigated. The work in this patch was originally authored by someone no longer at Arm

efriedma-quic commented on llvm/lib/Transforms/Scalar/ARMWidenStrings.cpp:
```diff
@@ -0,0 +1,227 @@
+// ARMWidenStrings.cpp - Widen strings to word boundaries to speed up
```
Missing license notification.

efriedma-quic commented on llvm/lib/Transforms/Scalar/ARMWidenStrings.cpp:
```diff
@@ -0,0 +1,227 @@
+      auto *IsVolatile = dyn_cast<ConstantInt>(CI->getArgOperand(3));
+
+      if (!BytesToCopy) {
+        LLVM_DEBUG(dbgs() << "Number of bytes to copy is null\n");
```
Consider cutting down the DEBUG lines; printing a message for every memcpy is extremely noisy (and not particularly useful without printing the instruction in question).

efriedma-quic reviewed (commented):
Probably worth investigating if we can fit this easily into some pass that's already examining the uses of globals, like GlobalOpt; iterating over the whole module isn't cheap.

nasherm committed:
Responding to review comments

Change-Id: I492ea4e5b6f589e5d877eeb6be31f7ab4720be9b

nasherm commented:
My most recent patch addresses the comments.

> Probably worth investigating if we can fit this easily into some pass that's already examining the uses of globals, like GlobalOpt; iterating over the whole module isn't cheap.

I've had a look at GlobalOpt briefly and I  have a few question: if this pass were added wouldn't investigation also include seeing if this improves performance on other targets? I can see restricting the pass to ARM cores but it seems like that would it make it a poor fit for GlobalOpt. Is there something I'm missing?

efriedma-quic commented:
If we're going to make this transform target-independent, we'll need some target-specific tuning from TargetTransformInfo or something like that.  Even if the transform is profitable, the exact thresholds where it's profitable are likely to be different.  (The maximum size of the global where it's relevant, and whether the best alignment boundary is 2/4/8/16 bytes, is going to vary.)

Not sure we need extensive performance measurements for other targets... if you could get measurements for some big x86 or Arm64 core, that would be nice.  But you can basically see what happens on other targets by just compiling a simple example.  And if we have a TTI hook, targets could easily opt-out.

nasherm committed:
Making ARMWidenStrings to be target independent

Change-Id: Ic6ed9a549e39020e8c04b38bc21ba8162b4ebfd9

nasherm commented:
I've rewritten the pass to be platform independent and added it to GlobalOpt. By default it's disabled for all targets except ARM.

nasherm commented:
I haven't had a chance to investigate performance for on AArch64 or x86 machines and will not be able to until next week

davemgreen commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,145 @@ OptimizeFunctions(Module &M,
+      unsigned int NumBytesToPad = 4 - (NumBytesToCopy % 4);
+      unsigned int TotalBytes = NumBytesToCopy + NumBytesToPad;
+
+      /*
```
Use //

davemgreen commented on llvm/test/Transforms/ARMWidenStrings/arm-widen-strings-1.ll:
```diff
@@ -0,0 +1,25 @@
+; RUN: opt < %s -mtriple=arm-arm-none-eabi -passes="default<O2>,globalopt" -S | FileCheck %s
```
It should be enough to test with -passes=globalopt, and shouldn't need the whole pass pipeline.

nasherm replied:
Done

nasherm committed:
Review comments

Updating patch so that when attempting to widen
global strings we only check whether the variable is
being called by a memcpy intrinsic.

Change-Id: I088403636c2ed0acc231af77b399b1b95f1abbc2

nasherm commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2058,6 +2199,14 @@ OptimizeGlobalVars(Module &M,
+    for (Function &F : llvm::make_early_inc_range(M)) {
+      TargetTransformInfo &TTI = GetTTI(F);
+      if (TTI.useWidenGlobalStrings()) {
+        Changed |= tryWidenGlobalStrings(F, GetTTI);
```
My latest patch checks the uses of a global variable for memcpys. I believe OptimizeFunctions seems to be primarily concerned with either changing a functions call convention or removing dead functions so I decided to keep the check in OptimizeGlobalVariables

nasherm commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,145 @@ OptimizeFunctions(Module &M,
+static bool
+tryWidenGlobalStrings(Function &F,
+                      function_ref<TargetTransformInfo &(Function &)> GetTTI) {
+  bool changed = false;
```
Done

davemgreen commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,129 @@ OptimizeFunctions(Module &M,
+  auto *BytesToCopy = dyn_cast<ConstantInt>(CI->getArgOperand(2));
+  auto *IsVolatile = dyn_cast<ConstantInt>(CI->getArgOperand(3));
+
+  if (!BytesToCopy)
```
if (!BytesToCopy || !Alloca || !Alloca->isStaticAlloca() || !IsVolatile || IsVolatile->isOne())

nasherm replied:
Done

davemgreen commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,129 @@ OptimizeFunctions(Module &M,
+  uint64_t DZSize = Alloca->getAllocatedType()->getArrayNumElements();
+  uint64_t SZSize = SourceDataArray->getType()->getNumElements();
+
+  // For safety purposes lets add a constraint and only padd when
```
padd -> pad

nasherm replied:
Done

davemgreen commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,129 @@ OptimizeFunctions(Module &M,
+
+  // Update destination char array to be word aligned (memcpy(X,...,...))
+  IRBuilder<> BuildAlloca(Alloca);
+  AllocaInst *NewAlloca = cast<AllocaInst>(BuildAlloca.CreateAlloca(
```
Can this cast be removed?

nasherm replied:
Done

davemgreen commented on llvm/test/Transforms/ARMWidenStrings/arm-widen-strings-struct-test.ll:
```diff
@@ -0,0 +1,52 @@
+; CHECK-NOT: [16 x i8]
+@.str = private unnamed_addr constant [13 x i8] c"hello world\0A\00", align 1
+@.str.1 = private unnamed_addr constant [4 x i8] c"%s\0A\00", align 1
+@__ARM_use_no_argv = global i32 1, section ".ARM.use_no_argv", align 4
```
Remove this one too, and generally try to clean up the tests. I would also run update_test_checks.py on them myself.

davemgreen replied:
They can also be moved to a llvm/test/Transforms/GlobalOpt now (maybe an ARM sub-folder if they are target specific).

nasherm replied:
Done

davemgreen commented on llvm/test/Transforms/ARMWidenStrings/arm-widen-strings-1.ll:
```diff
@@ -0,0 +1,25 @@
+
+declare i32 @bar(...) local_unnamed_addr
+
+; Function Attrs: mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite)
```
Remove some of the `Function Attrs`, `hidden`, `local_unnamed_addr`

davemgreen reviewed (commented):
Would it be possible to try and make the size a little more generic, not just 4? It could possibly have a cost-model based of the expected expansion of memcpy's, but that might not save much on many architectures other than the cost of an unaligned load/store.

Maybe the target could return the size to align to for a given array size, as opposed to the boolean useWidenGlobalStrings. So 4 for Arm (for ldm).

nasherm committed:
Review comments

nasherm commented:
> Maybe the target could return the size to align to for a given array size, as opposed to the boolean useWidenGlobalStrings. So 4 for Arm

Opted for this approach. Would it be worth making the chosen alignment value correspond to a target feature?

davemgreen commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,109 @@ OptimizeFunctions(Module &M,
+      Alloca->getAllocatedType()->getArrayElementType(), TotalBytes));
+  NewAlloca->takeName(Alloca);
+  NewAlloca->setAlignment(Alloca->getAlign());
+  Alloca->replaceAllUsesWith(NewAlloca);
```
Can Alloca be removed now?

nasherm replied:
Done

davemgreen reviewed (commented):
Thanks - I think this seems OK to me. I had some comments inline, but what currently happens when the global has multiple uses?

nasherm committed:
Responding to review comments

nasherm commented on llvm/test/Transforms/GlobalOpt/ARM/arm-widen-strings-2.ll:
```diff
@@ -0,0 +1,23 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -mtriple=arm-arm-none-eabi -passes="default<O2>,globalopt" -S | FileCheck %s
```
Done

nasherm commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,109 @@ OptimizeFunctions(Module &M,
+static bool IsCharArray(Type *T) {
+  const unsigned int CHAR_BIT_SIZE = 8;
+  return T && T->isArrayTy() && T->getArrayElementType()->isIntegerTy() &&
+         T->getArrayElementType()->getIntegerBitWidth() == CHAR_BIT_SIZE;
```
Yup. I've made the solution generic enough that it can theoretically pad any non-byte array. For ARM it will only attempt padding for arrays of integers. I think this *could* work for more complex types like structs but that might be out of the scope of the PR.

nasherm commented on llvm/include/llvm/Analysis/TargetTransformInfo.h:
```diff
@@ -1786,6 +1786,10 @@ class TargetTransformInfo {
 
+  /// \return For an array of given Size, return alignment boundary to
+  /// pad to. Default is no padding.
+  unsigned getNumBytesToPad(unsigned Size) const;
```
Done

nasherm commented on llvm/test/Transforms/GlobalOpt/ARM/arm-widen-strings-1.ll:
```diff
@@ -0,0 +1,33 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -mtriple=arm-arm-none-eabi -passes=globalopt -S | FileCheck %s
+; RUN: opt < %s -mtriple=arm-arm-none-eabi -passes="default<O0>" -S | FileCheck %s --check-prefix=TURNED-OFF
```
Done

davemgreen reviewed (commented):
> I've added an IR test that shows this. I believe the global variable only gets padded once while every destination array gets padded if using the variable

It looks like multiple globals get created. Can we prevent that from happening?

Can we replace all the uses of the old global with the new one, to make sure we don't end up with many repeated globals? i.e use the first n bytes of the now wider global in the "other" uses, and hopefully re-use the already widened global in other allocas that are also widened. (Or widen all uses at once, or maybe just limit it to a single use if it isn't very useful).

nasherm committed:
Review comments: eliminating generation of multiple globals

nasherm commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,115 @@ OptimizeFunctions(Module &M,
+  // division of integers returning wrong values i.e. copying one byte
+  // from an array of i16 would yield 0 elements to copy as supposed to 1.
+  unsigned NumElementsToCopy =
+      (NumBytesToCopy + ElementByteWidth - 1) / ElementByteWidth;
```
Yes, done

nasherm commented on llvm/test/Transforms/GlobalOpt/ARM/arm-widen-non-byte-array.ll:
```diff
@@ -0,0 +1,22 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -mtriple=arm-arm-none-eabi -passes=globalopt -S | FileCheck %s
```
Done

nasherm commented:
> Can we replace all the uses of the old global with the new one, to make sure we don't end up with many repeated globals? i.e use the first n bytes of the now wider global in the "other" uses, and hopefully re-use the already widened global in other allocas that are also widened. (Or widen all uses at once, or maybe just limit it to a single use if it isn't very useful)

I have something that works by not creating multiple globals. It rather creates the new global and updates all the references, but doesn't update the original source global. I'm not sure if this is an ideal solution.

nasherm committed:
Correcting and refactoring elimination

nasherm commented:
With the now updated pass I get the following output on the below (truncated) IR

**IR**
```
@.str = private unnamed_addr constant [10 x i8] c"123456789\00", align 1

; Function Attrs: nounwind
define hidden void @foo() #0 {
entry:
  %something = alloca [10 x i8], align 1
  call void @llvm.memcpy.p0.p0.i32(ptr noundef nonnull align 1 dereferenceable(10) %something, ptr noundef nonnull align 1 dereferenceable(10) @.str, i32 10, i1 false)
  %call2 = call i32 @bar(ptr nonnull %something)
  ret void
}


declare i32 @bar(...)
```

nasherm commented:
> With the now updated pass I get the following output on the below (truncated) IR
> 
> **IR**
> 
> ```
> @.str = private unnamed_addr constant [10 x i8] c"123456789\00", align 1
> 
> ; Function Attrs: nounwind
> define hidden void @foo() #0 {
> entry:
>   %something = alloca [10 x i8], align 1
>   call void @llvm.memcpy.p0.p0.i32(ptr noundef nonnull align 1 dereferenceable(10) %something, ptr noundef nonnull align 1 dereferenceable(10) @.str, i32 10, i1 false)
>   %call2 = call i32 @bar(ptr nonnull %something)
>   ret void
> }
> 
> 
> declare i32 @bar(...)
> ```

***ARM***
```
**ARM (No pass)** 
```
$ opt -mtriple=arm-none-eabi -S | llc  -o -
foo:
	.fnstart
@ %bb.0:                                @ %entry
	.save	{r11, lr}
	push	{r11, lr}
	.pad	#16
	sub	sp, sp, #16
	mov	r0, #57
	strh	r0, [sp, #12]
	ldr	r0, .LCPI0_0
	str	r0, [sp, #8]
	ldr	r0, .LCPI0_1
	str	r0, [sp, #4]
	add	r0, sp, #4
	bl	bar
	add	sp, sp, #16
	pop	{r11, lr}
	mov	pc, lr
	.p2align	2
@ %bb.1:
.LCPI0_0:
	.long	943142453                       @ 0x38373635
.LCPI0_1:
	.long	875770417                       @ 0x34333231
.Lfunc_end0:
	.size	foo, .Lfunc_end0-foo
	.fnend
                                        @ -- End function
	.type	.L.str,%object                  @ @.str
	.section	.rodata.str1.4,"aMS",%progbits,1
	.p2align	2, 0x0
.L.str:
	.asciz	"123456789"
	.size	.L.str, 10

	.section	".note.GNU-stack","",%progbits
	.eabi_attribute	30, 1	@ Tag_ABI_optimization_goals

```

**ARM (with pass)**
```
$ opt -passes=globalopt -mtriple=arm-none-eabi -S | llc  -o -
foo:
	.fnstart
@ %bb.0:                                @ %entry
	.save	{r11, lr}
	push	{r11, lr}
	.pad	#16
	sub	sp, sp, #16
	mov	r0, #57
	str	r0, [sp, #12]
	ldr	r0, .LCPI0_0
	str	r0, [sp, #8]
	ldr	r0, .LCPI0_1
	str	r0, [sp, #4]
	add	r0, sp, #4
	bl	bar
	add	sp, sp, #16
	pop	{r11, lr}
	mov	pc, lr
	.p2align	2
@ %bb.1:
.LCPI0_0:
	.long	943142453                       @ 0x38373635
.LCPI0_1:
	.long	875770417                       @ 0x34333231
.Lfunc_end0:
	.size	foo, .Lfunc_end0-foo
	.fnend
                                        @ -- End function
	.type	.L.str,%object                  @ @.str
	.section	.rodata,"a",%progbits
	.p2align	2, 0x0
.L.str:
	.asciz	"123456789\000\000"
	.size	.L.str, 12

	.section	".note.GNU-stack","",%progbits
	.eabi_attribute	30, 1	@ Tag_ABI_optimization_goals
```

nasherm commented:
> With the now updated pass I get the following output on the below (truncated) IR
> 
> **IR**
> 
> ```
> @.str = private unnamed_addr constant [10 x i8] c"123456789\00", align 1
> 
> ; Function Attrs: nounwind
> define hidden void @foo() #0 {
> entry:
>   %something = alloca [10 x i8], align 1
>   call void @llvm.memcpy.p0.p0.i32(ptr noundef nonnull align 1 dereferenceable(10) %something, ptr noundef nonnull align 1 dereferenceable(10) @.str, i32 10, i1 false)
>   %call2 = call i32 @bar(ptr nonnull %something)
>   ret void
> }
> 
> 
> declare i32 @bar(...)
> ```

***AArch64***
**AArch64 (no pass) **
```
	.text
	.file	"example.ll"
	.hidden	foo                             // -- Begin function foo
	.globl	foo
	.p2align	2
	.type	foo,@function
foo:                                    // @foo
	.cfi_startproc
// %bb.0:                               // %entry
	sub	sp, sp, #32
	str	x30, [sp, #16]                  // 8-byte Folded Spill
	.cfi_def_cfa_offset 32
	.cfi_offset w30, -16
	adrp	x9, .L.str
	add	x9, x9, :lo12:.L.str
	mov	w8, #57                         // =0x39
	ldr	x9, [x9]
	mov	x0, sp
	strh	w8, [sp, #8]
	str	x9, [sp]
	bl	bar
	ldr	x30, [sp, #16]                  // 8-byte Folded Reload
	add	sp, sp, #32
	ret
.Lfunc_end0:
	.size	foo, .Lfunc_end0-foo
	.cfi_endproc
                                        // -- End function
	.type	.L.str,@object                  // @.str
	.section	.rodata.str1.1,"aMS",@progbits,1
.L.str:
	.asciz	"123456789"
	.size	.L.str, 10

	.section	".note.GNU-stack","",@progbits
```

**AArch64 (with pass)**
```
$ opt -passes=globalopt -mtriple=aarch64-none-eabi | llc -o - 
foo:                                    // @foo
	.cfi_startproc
// %bb.0:                               // %entry
	sub	sp, sp, #32
	str	x30, [sp, #16]                  // 8-byte Folded Spill
	.cfi_def_cfa_offset 32
	.cfi_offset w30, -16
	adrp	x9, .L.str
	add	x9, x9, :lo12:.L.str
	mov	w8, #57                         // =0x39
	ldr	x9, [x9]
	mov	x0, sp
	str	w8, [sp, #8]
	str	x9, [sp]
	bl	bar
	ldr	x30, [sp, #16]                  // 8-byte Folded Reload
	add	sp, sp, #32
	ret
.Lfunc_end0:
	.size	foo, .Lfunc_end0-foo
	.cfi_endproc
                                        // -- End function
	.type	.L.str,@object                  // @.str
	.section	.rodata,"a",@progbits
.L.str:
	.asciz	"123456789\000\000"
	.size	.L.str, 12

	.section	".note.GNU-stack","",@progbits

```

nasherm committed:
Fix bug when copying to global dest

The case in which copying from a global
source to a global dest wasn't handled and
caused opt to crash. This is now handled and
a new test has been added to check

Change-Id: Ieb0467797fcee888f6e95e68af4dac9c05d70a4d

davemgreen commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2043,9 +2043,13 @@ static bool callInstIsMemcpy(CallInst *CI) {
 }
 
 static bool destArrayCanBeWidened(CallInst *CI) {
+  auto *GV = dyn_cast<GlobalVariable>(CI->getArgOperand(0));
```
getArgOperand(1)

nasherm replied:
Done

nasherm replied:
Removed as not necessary after reviewing

davemgreen commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,172 @@ OptimizeFunctions(Module &M,
+  // are memcpys.
+  for (auto *User : SourceVar->users()) {
+    auto *CI = dyn_cast<CallInst>(User);
+    if (!callInstIsMemcpy(CI))
```
Check that CI->getArgOperand(1) == GV here too? It should help rule out some cases we don't want to transform.

nasherm replied:
Done

nasherm committed:
Addressing review comments

Change-Id: I029312362f9dd714b2e9bc206cc002883d761b8b

nasherm committed:
Addressing review comments

Change-Id: Idc7b14cc785eb88552dd72947eb0df128baa7e90

nasherm commented on llvm/lib/Target/ARM/ARMTargetTransformInfo.cpp:
```diff
@@ -56,6 +56,10 @@ static cl::opt<bool>
     AllowWLSLoops("allow-arm-wlsloops", cl::Hidden, cl::init(true),
                   cl::desc("Enable the generation of WLS loops"));
 
+static cl::opt<unsigned> UseWidenGlobalStrings(
```
Done. Added to get padding function

nasherm committed:
Review comments

- Removed handling of global variable destinations. We simply
  don't pad these for now
- Added check that destination array is an array type and added
  test.

Change-Id: Ifc53051952ef69c4af64827402baf7d69cab4824

davemgreen reviewed (approved):
Thanks for the updates. I ran some tests and as far as I can tell they ran OK now. LGTM if there are no other comments.

davemgreen commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -2029,6 +2031,168 @@ OptimizeFunctions(Module &M,
+                                        const unsigned NumBytesToCopy,
+                                        ConstantInt *BytesToCopyOp,
+                                        ConstantDataArray *SourceDataArray) {
+  if (!SourceVar->hasInitializer() || !SourceVar->isConstant() ||
+      !SourceVar->hasLocalLinkage() || !SourceVar->hasGlobalUnnamedAddr())
+    return false;
```
Oh I forgot to mention - it would be good to have a test where the global isn't a constant.

nasherm committed:
Rebasing

Change-Id: Iad0539e526fb0fc116217dcbd033f8297fa5ef5f

nasherm committed:
Responding to review comments

- Added test showing behaviour of attempting to
  widen non-const globals
- Refactoring

Change-Id: I566214331bf3d889bd1409d3148aa6eab2530ed5

nasherm commented:
Has been reverted due to unexpected buildbot failures

aeubanks commented:
probably just requires `REQUIRES: arm-registered-target` in tests?
